[xbmc][music] Better handling of tags for multifile cue files

evilhamster opened the pull request:
As discussed in http://forum.kodi.tv/showthread.php?tid=189148 (a year ago, but the problem is still the same) if you have a cue file for your audio file(s), kodi will prefer the information from the cue file compared to the information in the tags. This is fine for cases when you have a single flac/cue (for example) but causes problems when you have multiple flac files with a cue and have re-tagged the flac files so that the tags are no longer in sync. This pr adds an advancedsetting that handles how cue files with multiple audiofiles should be handled, the default in this pr is to ignore information in cue files for these cases but that can be configured.

zag2me commented:
Massive +1 on this as I ran into this issue a few years ago. I ended up deleting all my cue files but i'm sure others will run into it.

fritsch commented:
I think fixing the "original" issue would be better.

E.g. check when the next file is entered and then update the meta information?

evilhamster commented:
I'm not sure I follow, must be slow today ;/

The problem is that if we have the following two albums (a and b, both have 2 songs each):
a-01.flac
a-02.flac
a.cue 
b.flac
b.cue
The way kodi currently works is that it uses the information in the cue file instead of the tags. This works fine for album b, since the tags on b.flac can't contain any information on the different tracks (unless a cue file is included in the tags). For album a this can cause problem (and have done so for me and others) since if we have updated the tags but not the .cue files, it will use the old information in the cue file. 

To fix this I can think of the following solutions:
1) You shouldn't have .cue files with wrong/old information so this should not be fixed.
2) Setting that tell Kodi to not use the information in the .cue files for album of type A but not type B
3) When fetching information from a .cue file, check the tags of that file and add them instead of the information in the .cue file if that file only contain one track. 

Since this problem has annoyed me, I didn't want to go with 1. 2 solves the problem and is an easy implemented solution. 3 would solve the problem as well but it would be more complex to implement and I don't see any benefit. 3 would also require some sort of setting telling it when to use the cue and when to try and use the tags. If you think this will cause problem, let me know what kind of problems you would like to avoid and I'll see what I can do.

btw, fixed the { } as requested.. old habits die hard :)

FernetMenta commented:
In most cases (like this one here) advanced settings are no real fixes. They are abused to adapt the behaviour to the needs of a few users only. The average joe isn't even aware of something like advanced settings. 

If read this right this change would break 1) (all .cue files are up to date) unless the user edits advanced settings. Correct?

Paxxi commented:
I fail to se why this is something we should have a setting for, especially as odd as this. If a user want the cue file then they should update it along with the tags.

There's no real way for us to solve this without lots of guessing and we're bound to guess wrong sometimes.

evilhamster commented:
The problem is that you have metadata on two places (both in the cue and in the flacs), one has to be preferred to the other. This isn't a problem that you fix, but something you try and make the best of the situation. I think that adapting this makes for a better experience for the average user, but I could be wrong.  

It will not break for all .cue files. The only case it will break is when using a compliant .cue that contains multiple audiofiles that do not contain tags or contain tags with old information (a common version of .cue files with multiple audiofiles are non-compliant, http://wiki.hydrogenaud.io/index.php?title=EAC_and_Cue_Sheets, and not handled by xbmc).

As it currently stands:
cue/single flac -> information from the cue.
cue/multiple flac (compliant cue) -> information from the cue.
cue/multiple flac (non-compliant cue) -> information from the tags.

Using this pr:
cue/single flac -> information from the cue.
cue/multiple flac (compliant cue) -> information from the tags.
cue/multiple flac (non-compliant cue) -> information from the tags.

The tagging tools I'm familiar with does not update the .cue file but only the .flac's. If we keep it as is, users will use a tagging tool and be surprised when it is not visible in Kodi (unless they where lucky and the cue file was non-compliant). Forcing users to go the way of zag2me and wipe all .cue files does not sound like a user friendly approach. If some advanced users wish to use the information in the .cue files for all cases, they can update the advancedsettings. But for the rest of us, it should pickup the information from the most likely places and for multiple-files that should imho be the tags. 

As Paxxi points out, if you would like to have the .cue files they should be updated. The problem is that if you would like to tag your music with mbrainztags, there is no standard way of saving that information in the cue (afaik). Or did I misunderstand and you are just against having this as an advanced setting?

Paxxi commented:
Thanks for the explanation @evilhamster . Is there any scenario where a cue file would be preferable in a multi file situation and provide any benefit?

As you describe the situation my idea would be, no setting, for multiple files, read id3tag, if none available consult cue file? Would that work?

zag2me commented:
In my opinion the file tags should always be primary for single tracks, and the cue primary source for single flac/mp3 full CD's. Having run into this problem before, the new behaviour of this PR is exactly how I would have expected it to work.

I'm not even sure it needs the advanced setting, but it does give us a fall-back for any users who wish to use cue files (there is a scene of audiophile people who rip tracks to lossless with exact audio copy ripping tools which this setting may benefit)

fritsch commented:
Yeah, I know this scene - most of the time they are whining when they don't
see the green light on their 3000 dollars DAC ...

So, when the PR does the following:

if (multiple tracks are seen and track has tag information) use track's tag
information
if (multiple tracks are seen but no tag information given) use cues track

Then there does not need to be a setting at all.

2015-06-16 10:09 GMT+02:00 Zag <notifications@github.com>:

> In my opinion the file tags should always be primary for single tracks,
> and the cue primary source for single flac/mp3 full CD's. Having run into
> this problem before, the new behaviour of this PR is exactly how I would
> have expected it to work.
>
> I'm not even sure it needs the advanced setting, but it does give us a
> fall-back for any users who wish to use cue files (there is a scene of
> audiophile people who rip tracks to lossless with exact audio copy ripping
> tools which this setting may benefit)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7282#issuecomment-112326846>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

FernetMenta commented:
instead of a setting we could add and parse a comment in the .cue file:
REM kodi please ignore

fritsch commented:
Yeah, let's parse the REM comment:
https://en.wikipedia.org/wiki/Cue_sheet_(computing)

That way, we don't introduce a kodi only standard, e.g.:

```
REM KODI_IGNORE_CUE_METADATA_01
```

tells kodi to ignore Cue's metadata for track1 and read the flac track meta
data byself.

2015-06-16 11:11 GMT+02:00 Rainer Hochecker <notifications@github.com>:

> instead of a setting we could add and parse a comment in the .cue file:
> REM kodi please ignore
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7282#issuecomment-112351947>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

zag2me commented:
Wouldn't that mean you would have to edit every single CUE file?

I have 5,000 CD's ;)

But yeh maybe the other way round "REM kodi please read cue metadata"

fritsch commented:
I still don't get at all, why someone saves cue sheets with wrong
information! A cue sheet is a container with meta information - just
automatically remove all the tags if you don't want them ...

2015-06-16 11:47 GMT+02:00 Zag <notifications@github.com>:

> Wouldn't that mean you would have to edit every single CUE file?
>
> I have 5,000 CD's ;)
>
> But yeh maybe the other way round "REM kodi please read cue metadata"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7282#issuecomment-112367496>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

zag2me commented:
EAC (http://www.exactaudiocopy.de/) which is the most popular ripper software, saves the cue sheets automatically so the user has no input in it. It automatically looks up the CD info on some on-line sites which are less accurate than musicbrainz.

fritsch commented:
Wait - they have an issue with being accurate, when ripping from a bought CD?

zag2me commented:
Yes the name tagging can be wrong as it uses external database apis

evilhamster commented:
@Paxxi 
Users that keep their audio files as wav and use .cue files for the tag information and used a compliant way to save the gap information (most guides I have seen seem to favor the non-compliant way of saving that information, if you are interested there is some good information at http://wiki.hydrogenaud.io/index.php?title=EAC_Gap_Settings).

It would be possible to do something like that, but there are drawbacks.

You could update the ParseFile function for .cue files to check the tags on the files and if found use the tags on the files instead of the information in the cue. This would have the drawback that everytime you entered a directory (in file mode) with a .cue file it would try and read the tags (as apposed to the first time).

You could update the LoadTracksFromCueDocument function and force it to use the tags instead of the cue information for these cases (if the exist). It would also require update the cuedocument class with a flag if it contain multiple files or not. Apart from being more lines of code, I don't really see any other drawbacks. 

If the second option would be ok to get merged, I'll update this PR (or should I create a new one?).  

@fritsch 
Some of the information between the cue file and the tags overlap, but some information is only found in the .cue file (for exampel discid, isrc, gap information) and some information is only found in the tags (for example musicbrainz information).

One real world example.

You buy a cd like http://musicbrainz.org/release/43abbb33-c968-476b-9376-1b22756109ac (a album that is made of several artists). You decide to rip it to flac (as multiple files) with a cue file (since you would like to keep gap information if you would like to burn it later on) and add it to Kodi. You will now have the artist "Afel Bocoum, Damon Albarn, Toumani Diabaté, Ko Kan Ko Sata" added to your artist list. To sort this you fire up picard and add the correct mbrainz tags to the flac files (since there is no defined way of saving the mbrainz information in the .cue file).  To get kodi to pickup these changes you have the following options:
* Nuke the .cue files (but then you will lose some information as previously pointed out)
* Rename the .cue files to something else.
* Patch kodi so that it dont prefer the information in the .cue file. 

Since I don't see any downsides with the last point and the two first one causes unnecessary work I think it should be referred :)

Paxxi commented:
@evilhamster I'm good with the second solution you mentioned, sounds like a good way to handle it.

evilhamster committed:
Prefer information from tags instead of cue files if there are one audiofile for each track and the tags contain information about artist, album and title.

evilhamster commented:
Update the code as request. The name of the PR is no longer valid. 

The way it works now is that CueDocument has a bool variable that gets set true if the number of tracks in a cue file matches the number of tracks. If this variable is set and the tags on the file contains artist, title and album it will use the tags instead of the information from the cue.

Paxxi commented:
@evilhamster I took the liberty to update the PR name, feel free to change it if you like :)

Paxxi commented:
@evilhamster code looks good, +1 from me.

fritsch commented:
Code looks in deed niceley done. This:
https://github.com/evilhamster/xbmc/commit/9ec1eefbd35b9ba74dac7df29c7c9124a8a20803#diff-6db2f8726d6229b9edcb6ceef3172652R82
is semantically a bit problematic, but yeah - not your fault.

If all are happy +1 from me. Thanks much.

2015-06-19 10:47 GMT+02:00 Pär Björklund <notifications@github.com>:

> @evilhamster <https://github.com/evilhamster> code looks good, +1 from me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7282#issuecomment-113435590>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

MartijnKaijser commented:
jenkins build this please
any objections on merging?

MartijnKaijser commented:
@fritsch please merge when ok for you
